### PR TITLE
Wire J2CL composer toolbar attachments

### PR DESCRIPTION
**Pull request by vega113**
## Summary

Refs #971 and #904.

Implements #971 Task 5 by wiring the J2CL/Lit composer and toolbar through the existing #969 surface instead of creating a parallel composer. The root shell now routes toolbar actions to `J2clComposeSurfaceController`, uses the rich-content delta factory, and creates the attachment composer with the root-shell session seed.

This PR adds the Task 5 toolbar subset for attachments, keeps future rich/attachment enum IDs stable but unsurfaced, surfaces command status/error through Lit live regions, supports file-picker and pasted-image uploads, and submits uploaded attachments as structured rich reply content only after upload success. It also guards no-wave, sign-out, cancel, late callback, failed submit, and in-flight submit races.

## Review Loop

External Claude Opus 4.7 review was run repeatedly through round 11. Follow-up fixes from the loop include picker-cancel quiet state, no-wave active-command normalization, mixed text+image paste preservation, pending-upload submit guard, reply-submitting attachment entry-point block, rich-command toggle-off, cancel semantics for already-inserted attachments, neutral signed-out copy, Lit listener cleanup/reconnect coverage, and hidden picker form detachment.

## Verification

Recorded locally in `journal/local-verification/2026-04-25-issue-971-composer-toolbar-wiring.md` and mirrored here:

- `cd j2cl && ./mvnw -Dtest="J2clComposeSurfaceControllerTest#cancelAttachmentToolbarActionKeepsAlreadyInsertedAttachments,J2clComposeSurfaceControllerTest#cancelAttachmentToolbarActionClearsUploadQueueState" test` — passed, 2 tests.
- `cd j2cl/lit && npm test -- test/composer-inline-reply.test.js` — passed, 13 tests.
- `cd j2cl && ./mvnw -Dtest="J2clToolbarSurfaceControllerTest,J2clComposeSurfaceControllerTest,J2clRichContentDeltaFactoryTest,J2clAttachmentComposerControllerTest,J2clAttachmentUploadClientTest,J2clAttachmentIdGeneratorTest" test` — passed, 156 tests.
- `cd j2cl/lit && npm test && npm run build` — passed, 21 files / 96 tests; build passed.
- `cd j2cl && ./mvnw test` — passed, 499 tests, 61 skipped.
- `python3 scripts/assemble-changelog.py && python3 scripts/validate-changelog.py --fragments-dir wave/config/changelog.d --changelog wave/config/changelog.json` — passed.
- `git diff --check` and `git diff --cached --check` — passed.

## Summary by CodeRabbit

* **New Features**
  * Rich-text toolbar (headings, indent/outdent) and expanded attachment flows: file picker, paste-image uploads, upload queue, cancel, size selection, captions/open/download, attachment-only replies; live command status/error announcements and composer focus restoration; reply submission now waits for uploads to finish.

* **Tests**
  * Broad test coverage added for toolbar, attachment picker/paste, upload lifecycle, focus, and draft/submit edge cases.

* **Documentation**
  * Changelog entry added for toolbar/attachment wiring.

### CODE REVIEW

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java [hunk 1]
```diff
@@ -372,7 +665,323 @@ private void render() {
...
+    String annotationKey = annotationKey(action);
+    String annotationValue = annotationValue(action);
+    if (annotationKey != null
+        && annotationValue != null
+        && draftText != null
+        && !draftText.isEmpty()) {
+      builder.annotatedText(annotationKey, annotationValue, draftText);
+    } else {
+      builder.text(draftText);
+    }
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java:729]: Fixed. Changed to `!draftText.trim().isEmpty()` so whitespace-only drafts fall through to `builder.text()` instead of triggering `annotatedText()`.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java [hunk 2]
```diff
@@ -372,7 +665,323 @@ private void render() {
...
+      return;
+    }
+    if (action == J2clDailyToolbarAction.ATTACHMENT_UPLOAD_QUEUE) {
+      refreshAttachmentCommandState();
+      if (commandStatusText.isEmpty() && commandErrorText.isEmpty()) {
+        commandStatusText = "No attachment uploads are queued.";
+      }
+      render();
+      view.focusReplyComposer();
+      return;
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java:868]: Fixed. Now clears `activeCommandId`, `commandStatusText`, and `commandErrorText` before calling `refreshAttachmentCommandState()`, so stale formatting status can't suppress the empty-queue fallback message.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/attachment/J2clAttachmentComposerController.java [hunk 3]
```diff
@@ -304,11 +321,13 @@ private void startNextUpload() {
   private void startUpload(QueueItem item) {
     uploadInProgress = true;
     item.status = UploadStatus.UPLOADING;
+    notifyStateChanged();
     int generation = resetGeneration;
     J2clAttachmentUploadClient.UploadProgressCallback progressCallback =
         percent -> {
           if (generation == resetGeneration && item.status == UploadStatus.UPLOADING) {
             item.progressPercent = clampPercent(percent);
+            notifyStateChanged();
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/attachment/J2clAttachmentComposerController.java]: Fixed. Progress callback now compares clamped percent before updating and only calls `notifyStateChanged()` when the value actually changes.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java [hunk 4]
```diff
@@ -154,10 +288,137 @@ public void onReplySubmitted(String draft) {
...
+      List<J2clAttachmentComposerController.AttachmentSelection> attachmentSelections =
+          new ArrayList<J2clAttachmentComposerController.AttachmentSelection>();
+      for (AttachmentFileSelection selection : selections) {
+        requirePresent(selection, "Attachment selection is required.");
+        attachmentSelections.add(
+            J2clAttachmentComposerController.AttachmentSelection.file(
+                selection.getPayload(), selection.getFileName(), "", attachmentDisplaySize));
+      }
+      activeCommandId = J2clDailyToolbarAction.ATTACHMENT_UPLOAD_QUEUE.id();
+      ensureAttachmentController().selectFiles(attachmentSelections);
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java:404]: False positive — the controller separates `activeCommandId` (UI display state) from `annotationCommandId` (annotation applied to the document). Attachment actions only overwrite `activeCommandId`. `buildDocument()` reads `annotationCommandId` at line 684, which is never touched during attachment selection, so bold/formatting applied before queueing attachments is fully preserved on submit. Added `richToolbarAnnotationSurvivesAttachmentUploadErrorStatus` test to cover this explicitly.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java [hunk 5]
```diff
@@ -154,10 +289,141 @@ public void onReplySubmitted(String draft) {
...
+      ensureAttachmentController().pasteImage(imagePayload, "", attachmentDisplaySize);
+      refreshAttachmentCommandState();
+    } catch (RuntimeException e) {
+      activeCommandId = ATTACHMENT_ERROR_STATE_ID;
+      commandStatusText = "";
+      commandErrorText = messageOrDefault(e, "Pasted image upload failed.");
+    }
+    render();
+    view.focusReplyComposer();
+  }
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java:449]: Fixed. Added `signedOut` guard at the top of both `onAttachmentFilesSelected` and `onPastedImage`, producing 'Sign in before attaching files.' / 'Sign in before pasting an image.' instead of falling through to the no-wave guard.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java [hunk 6]
```diff
@@ -372,7 +673,327 @@ private void render() {
...
+    activeCommandId = J2clDailyToolbarAction.ATTACHMENT_UPLOAD_QUEUE.id();
+    commandStatusText = "Attached " + insertion.getCaption() + ".";
+    commandErrorText = "";
+    render();
+  }
+
+  private void onAttachmentStateChanged() {
+    refreshAttachmentCommandState();
+    render();
+  }
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java:921]: Fixed. `onAttachmentInserted` now clears `replyErrorText` unconditionally. `onAttachmentStateChanged` clears it when `hasPendingAttachmentUpload()` returns false and the error starts with 'Wait for attachment', so only the upload-blocking message is cleared — not unrelated reply errors.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/attachment/J2clAttachmentComposerController.java [hunk 7]
```diff
@@ -401,6 +421,10 @@ private void insertAttachment(QueueItem item) {
     insertionCallback.onInsert(document, insertion);
   }
 
+  private void notifyStateChanged() {
+    stateChangeCallback.onStateChanged();
+  }
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/attachment/J2clAttachmentComposerController.java:437]: Already fixed — notifyStateChanged wraps the callback in try/catch RuntimeException in the current code.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java [hunk 8]
```diff
@@ -349,6 +646,10 @@ private void handleReplyFailure(int generation, String error) {
     replySubmitting = false;
     replyStatusText = "";
     replyErrorText = error == null || error.isEmpty() ? "Reply failed." : error;
+    activeCommandId = "";
+    annotationCommandId = "";
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java]: Fixed in f23dc8888 — handleReplyFailure no longer clears annotationCommandId, so retries carry the original Bold/Italic formatting.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java [hunk 9]
```diff
@@ -219,9 +506,11 @@ private void submitCreate() {
           if (generation != createGeneration) {
             return;
           }
-          J2clPlainTextDeltaFactory.CreateWaveRequest request;
+          CreateWaveRequest request;
           try {
-            request = deltaFactory.createWaveRequest(bootstrap.getAddress(), submittedDraft);
+            request =
+                deltaFactory.createWaveRequest(
+                    bootstrap.getAddress(), submittedDraft, buildDocument(submittedDraft, false));
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java]: Fixed in 3a47d9cd — buildDocument now resolves the annotation only on the reply path (includeAttachments=true). For create-wave (includeAttachments=false), action is set to null so no reply-toolbar annotation leaks into newly created waves.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java:729]: Fixed in 3a47d9cd — resetAttachmentState is now guarded by nextWaveId != null, so a transient null-session reconnect no longer drops in-flight attachments.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java:868]: Fixed in f23dc8888 — handleReplyFailure no longer resets annotationCommandId, preserving formatting for retries.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java [hunk 10]
```diff
@@ -329,6 +642,18 @@ private void handleReplyResponse(
...
+    insertedAttachments.clear();
+    // A sent reply closes the attachment batch; failures preserve size and attachments for retry.
+    attachmentDisplaySize = J2clAttachmentComposerController.DisplaySize.MEDIUM;
+    activeCommandId = "";
+    annotationCommandId = "";
+    commandStatusText = "";
+    commandErrorText = "";
+    J2clAttachmentComposerController completedAttachmentController = attachmentController;
+    attachmentController = null;
+    if (completedAttachmentController != null) {
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java]: Fixed in f1ec20a7e. The controller is no longer nulled after a successful submit — `cancelAndReset()` is called instead, which preserves the id generator counter (the javadoc states: "The attachment id generator is intentionally not reset"). Since `ensureAttachmentController()` returns the same instance for the next batch, the counter continues from where it left off. Test: `attachmentIdsContinueAcrossSuccessfulReplyBatches`.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java [hunk 11]
```diff
@@ -190,6 +479,14 @@ public void onWriteSessionChanged(J2clSidecarWriteSession nextWriteSession) {
         }
       }
     }
+    if (waveChanged) {
+      String nextWaveId = nextWriteSession == null ? null : nextWriteSession.getSelectedWaveId();
+      // Only reset attachments when transitioning to a real, different wave.
+      // A null nextWriteSession is a transient disconnect — preserve in-flight attachments.
+      if (nextWaveId != null) {
+        resetAttachmentState();
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java]: False positive. `selectedWaveChanged()` already handles this correctly. It returns `false` when `nextWriteSession == null` (disconnect), so the `if (waveChanged)` block is never entered during a disconnect. On reconnect to the same wave, `currentWaveId` is read from `lastSelectedWaveId` (which persists through the null transition), so `safeEquals(currentWaveId, nextWaveId)` is true and `selectedWaveChanged` again returns `false` — attachment state is not reset. Tests `sameWaveReconnectPreservesInFlightAttachmentUpload` and `differentWaveReconnectAfterDisconnectDropsInFlightAttachmentUpload` cover both paths.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java]: Fixed in f1ec20a7e. Successful reply batches now keep the attachment controller lifecycle state instead of nulling it, so the attachment id counter continues across batches; attachmentIdsContinueAcrossSuccessfulReplyBatches covers seedA then seedB. Verified with sbt -batch j2clSearchTest.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java]: Fixed in f1ec20a7e. The controller now remembers the last selected wave across null write-session disconnects, preserving in-flight attachments on same-wave reconnect while resetting them on different-wave reconnect. Covered by sameWaveReconnectPreservesInFlightAttachmentUpload and differentWaveReconnectAfterDisconnectDropsInFlightAttachmentUpload. Verified with sbt -batch j2clSearchTest.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java [hunk 12]
```diff
@@ -372,7 +701,350 @@ private void render() {
...
+      attachmentDisplaySize = J2clAttachmentComposerController.DisplaySize.LARGE;
+      commandStatusText = "Large attachment size selected.";
+      render();
+      view.focusReplyComposer();
+      return;
+    }
+    if (action == J2clDailyToolbarAction.ATTACHMENT_CANCEL) {
+      if (attachmentController != null) {
+        attachmentController.cancelAndReset();
+        attachmentController = null;
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java]: Fixed in 7de1556bd. ATTACHMENT_CANCEL now calls cancelAndReset without discarding attachmentController, preserving the id generator for subsequent attachments in the same reply batch. Extended cancelAttachmentToolbarActionPreservesInsertedAndContinuesIdGenerator and added cancelAttachmentToolbarActionPreservesIdGeneratorAfterPendingUploadCancel to cover seedA then seedB after cancel. Verified with sbt -batch j2clSearchTest.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java [hunk 13]
```diff
@@ -154,10 +298,157 @@ public void onReplySubmitted(String draft) {
...
+      view.focusReplyComposer();
+      return true;
+    }
+    if (safeEquals(annotationCommandId, action.id())) {
+      activeCommandId = "";
+      annotationCommandId = "";
+      commandErrorText = "";
+      commandStatusText = action.label() + " cleared.";
+      render();
+      view.focusReplyComposer();
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java:353]: Fixed in commit 7de1556bd. The toggle-off branch now mirrors CLEAR_FORMATTING: commandStatusText is cleared first, then refreshAttachmentCommandState() runs, and 'X cleared.' is only set if both fields remain empty. Regression tests added for uploading and error states.

File: j2cl/lit/src/elements/composer-inline-reply.js [hunk 14]
```diff
@@ -91,6 +127,40 @@ export class ComposerInlineReply extends LitElement {
   onSubmit() {
     this.dispatchEvent(new CustomEvent("reply-submit", { bubbles: true, composed: true }));
   }
+
+  onPaste(event) {
+    const items = Array.from(event.clipboardData?.items || []);
+    const file = items
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/lit/src/elements/composer-inline-reply.js]: Already fixed in 8b0c4911. The `onPaste` handler now falls back to `clipboardData.files` when `clipboardData.items` yields no image — lines 138-141 in the current file. Verified the code is in.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java [hunk 15]
```diff
@@ -372,7 +705,349 @@ private void render() {
...
+  }
+
+  private J2clComposerDocument buildDocument(
+      String draftText, boolean includeAttachments, boolean includeAnnotation) {
+    J2clComposerDocument.Builder builder = J2clComposerDocument.builder();
+    // Only apply the reply-toolbar annotation on the reply path; create-wave submits must not
+    // inherit a Bold/Italic toggle the user set in the reply editor.
+    J2clDailyToolbarAction action =
+        includeAnnotation ? J2clDailyToolbarAction.fromId(annotationCommandId) : null;
+    String annotationKey = annotationKey(action);
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java]: Fixed in 452d2b22e. Captured `annotationCommandId` into `final String submittedAnnotationCommandId` at submit time (alongside `submittedDraft`), and passed it through a refactored `buildDocument(draftText, includeAttachments, annotationCommandId)` that takes the annotation ID as a `String` directly (empty = no annotation). Added regression test `replySubmitUsesFormattingSnapshotFromSubmitClick`: applies Bold, submits, toggles Italic before bootstrap resolves, then asserts the submitted delta uses Bold (not Italic).

[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java]: Fixed in 452d2b22e with wording cleanup in 33abb9c61. submitReply snapshots annotationCommandId before async bootstrap and passes that captured id into buildDocument; replySubmitUsesFormattingSnapshotFromSubmitClick covers Bold at send time followed by Italic while bootstrap is pending, with the submitted delta still using bold and no fontStyle. Verified with sbt -batch j2clSearchTest.

File: j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java [hunk 16]
```diff
@@ -372,7 +706,347 @@ private void render() {
...
+  public static DeltaFactory richContentDeltaFactory(String sessionSeed) {
+    return richContentDeltaFactory(new J2clRichContentDeltaFactory(sessionSeed));
+  }
+
+  public static AttachmentControllerFactory attachmentControllerFactory(String sessionSeed) {
+    return (waveRef, domain, insertionCallback, stateChangeCallback) ->
+        new J2clAttachmentComposerController(
+            waveRef,
+            new J2clAttachmentUploadClient(),
+            new J2clAttachmentIdGenerator(domain, sessionSeed),
```
[the remark this answers is not in the document]
Comment (vega113) [reply j2cl/src/main/java/org/waveprotocol/box/j2cl/compose/J2clComposeSurfaceController.java:750]: False positive. The `attachmentControllerFactory` captures `idGeneratorsByDomain` as a closure-level `Map<String, J2clAttachmentIdGenerator>` (lines 755-756). `attachmentIdGeneratorForDomain()` looks up the map on every `create()` call and only allocates a new generator when the domain is unseen. Since the factory itself is never replaced across wave transitions — only the controller instance is nulled in `resetAttachmentState()` — a new controller created for a different wave in the same domain always receives the same `J2clAttachmentIdGenerator` instance, so the counter continues from where it left off. IDs are monotonic within a domain for the full lifetime of the surface controller.
